From patchwork Tue Aug 27 16:04:02 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Vitaly Kuznetsov <vkuznets@redhat.com>
X-Patchwork-Id: 11117239
Return-Path: <SRS0=KFyJ=WX=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id BE55A1395
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 27 Aug 2019 16:04:32 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id A67012173E
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 27 Aug 2019 16:04:32 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1730017AbfH0QEM (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 27 Aug 2019 12:04:12 -0400
Received: from mx1.redhat.com ([209.132.183.28]:42056 "EHLO mx1.redhat.com"
        rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
        id S1729671AbfH0QEL (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 27 Aug 2019 12:04:11 -0400
Received: from mail-wm1-f72.google.com (mail-wm1-f72.google.com
 [209.85.128.72])
        (using TLSv1.2 with cipher ECDHE-RSA-AES128-GCM-SHA256 (128/128 bits))
        (No client certificate requested)
        by mx1.redhat.com (Postfix) with ESMTPS id 888332A09AB
        for <kvm@vger.kernel.org>; Tue, 27 Aug 2019 16:04:10 +0000 (UTC)
Received: by mail-wm1-f72.google.com with SMTP id u21so1251026wml.4
        for <kvm@vger.kernel.org>; Tue, 27 Aug 2019 09:04:10 -0700 (PDT)
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=VhVONP1yOZ9qqPvIKZXfSIL1EWyqy2BDr9yNoXDMCHw=;
        b=EHmNeSx45OpPCG7zpjVVgC1B2inGdu+3oGC2JPh3hyfJi+L04QQLzM4IBLmOCdVFZe
         SqQ6yEHYtLuBF5BOyBWRH2O0u590Bj9YVwT0t3wlp9+UyhVz3p/xxxKwI4OZt6Lzp8P1
         V4uXa2qcO3lI4HvLs0GebY0QyVzuXmRT0mkxVygjU+NgDWjOZ3sUEvErkBhB7UA89Ykt
         V0oHCenXPWwNa0idH5u5O4hQ9B8IATJrCvF1o7dVQVAzUXuODUGNrpfj/CUas42z7fjh
         ludOikRpR3WuayC7YHVoEYODccqLPrYN+wGix1W+wI44rUhliTaWhu3RKWVgWNfBtxan
         Vduw==
X-Gm-Message-State: APjAAAXTw7ajOFzceHIV7CP51Kub2ROa18T/+SWb0lv/CYCaSbhpeYCE
        O1MjVqYt7dOSJyo4FEpiBDse+d7RcCLvBheCvLg8K9+lHedFHFK1B7rx+ltWyJqIASIRLhWghlp
        oae4MhwjNdtlR
X-Received: by 2002:a1c:f110:: with SMTP id p16mr28612684wmh.59.1566921848499;
        Tue, 27 Aug 2019 09:04:08 -0700 (PDT)
X-Google-Smtp-Source: 
 APXvYqwQuwukGsraHaatE4hAj0LDnedGOIKHMZOGMIdh4CEFhBseJtI9ADNJhCCUprGZX99+Uaeilg==
X-Received: by 2002:a1c:f110:: with SMTP id p16mr28612666wmh.59.1566921848295;
        Tue, 27 Aug 2019 09:04:08 -0700 (PDT)
Received: from vitty.brq.redhat.com (nat-pool-brq-t.redhat.com.
 [213.175.37.10])
        by smtp.gmail.com with ESMTPSA id
 n8sm13461246wro.89.2019.08.27.09.04.07
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Tue, 27 Aug 2019 09:04:07 -0700 (PDT)
From: Vitaly Kuznetsov <vkuznets@redhat.com>
To: kvm@vger.kernel.org
Cc: linux-kernel@vger.kernel.org, Paolo Bonzini <pbonzini@redhat.com>,
	=?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Joerg Roedel <joro@8bytes.org>, Jim Mattson <jmattson@google.com>,
 Sean Christopherson <sean.j.christopherson@intel.com>,
 Roman Kagan <rkagan@virtuozzo.com>
Subject: [PATCH 1/3] KVM: x86: hyper-v: don't crash on
 KVM_GET_SUPPORTED_HV_CPUID when kvm_intel.nested is disabled
Date: Tue, 27 Aug 2019 18:04:02 +0200
Message-Id: <20190827160404.14098-2-vkuznets@redhat.com>
X-Mailer: git-send-email 2.20.1
In-Reply-To: <20190827160404.14098-1-vkuznets@redhat.com>
References: <20190827160404.14098-1-vkuznets@redhat.com>
MIME-Version: 1.0
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

If kvm_intel is loaded with nested=0 parameter an attempt to perform
KVM_GET_SUPPORTED_HV_CPUID results in OOPS as nested_get_evmcs_version hook
in kvm_x86_ops is NULL (we assign it in nested_vmx_hardware_setup() and
this only happens in case nested is enabled).

Check that kvm_x86_ops->nested_get_evmcs_version is not NULL before
calling it. With this, we can remove the stub from svm as it is no
longer needed.

Fixes: e2e871ab2f02 ("x86/kvm/hyper-v: Introduce nested_get_evmcs_version() helper")
Signed-off-by: Vitaly Kuznetsov <vkuznets@redhat.com>
Reviewed-by: Jim Mattson <jmattson@google.com>
---
 arch/x86/kvm/hyperv.c  | 5 ++++-
 arch/x86/kvm/svm.c     | 8 +-------
 arch/x86/kvm/vmx/vmx.c | 1 +
 3 files changed, 6 insertions(+), 8 deletions(-)

diff --git a/arch/x86/kvm/hyperv.c b/arch/x86/kvm/hyperv.c
index 58bf61b17431..3f5ad84853fb 100644
--- a/arch/x86/kvm/hyperv.c
+++ b/arch/x86/kvm/hyperv.c
@@ -1785,7 +1785,7 @@ int kvm_vm_ioctl_hv_eventfd(struct kvm *kvm, struct kvm_hyperv_eventfd *args)
 int kvm_vcpu_ioctl_get_hv_cpuid(struct kvm_vcpu *vcpu, struct kvm_cpuid2 *cpuid,
 				struct kvm_cpuid_entry2 __user *entries)
 {
-	uint16_t evmcs_ver = kvm_x86_ops->nested_get_evmcs_version(vcpu);
+	uint16_t evmcs_ver = 0;
 	struct kvm_cpuid_entry2 cpuid_entries[] = {
 		{ .function = HYPERV_CPUID_VENDOR_AND_MAX_FUNCTIONS },
 		{ .function = HYPERV_CPUID_INTERFACE },
@@ -1797,6 +1797,9 @@ int kvm_vcpu_ioctl_get_hv_cpuid(struct kvm_vcpu *vcpu, struct kvm_cpuid2 *cpuid,
 	};
 	int i, nent = ARRAY_SIZE(cpuid_entries);
 
+	if (kvm_x86_ops->nested_get_evmcs_version)
+		evmcs_ver = kvm_x86_ops->nested_get_evmcs_version(vcpu);
+
 	/* Skip NESTED_FEATURES if eVMCS is not supported */
 	if (!evmcs_ver)
 		--nent;
diff --git a/arch/x86/kvm/svm.c b/arch/x86/kvm/svm.c
index 572c7c4ca974..40175c42f116 100644
--- a/arch/x86/kvm/svm.c
+++ b/arch/x86/kvm/svm.c
@@ -7122,12 +7122,6 @@ static int svm_unregister_enc_region(struct kvm *kvm,
 	return ret;
 }
 
-static uint16_t nested_get_evmcs_version(struct kvm_vcpu *vcpu)
-{
-	/* Not supported */
-	return 0;
-}
-
 static int nested_enable_evmcs(struct kvm_vcpu *vcpu,
 				   uint16_t *vmcs_version)
 {
@@ -7344,7 +7338,7 @@ static struct kvm_x86_ops svm_x86_ops __ro_after_init = {
 	.mem_enc_unreg_region = svm_unregister_enc_region,
 
 	.nested_enable_evmcs = nested_enable_evmcs,
-	.nested_get_evmcs_version = nested_get_evmcs_version,
+	.nested_get_evmcs_version = NULL,
 
 	.need_emulation_on_page_fault = svm_need_emulation_on_page_fault,
 
diff --git a/arch/x86/kvm/vmx/vmx.c b/arch/x86/kvm/vmx/vmx.c
index 3c936e7366b9..c81e5210b159 100644
--- a/arch/x86/kvm/vmx/vmx.c
+++ b/arch/x86/kvm/vmx/vmx.c
@@ -7812,6 +7812,7 @@ static struct kvm_x86_ops vmx_x86_ops __ro_after_init = {
 	.set_nested_state = NULL,
 	.get_vmcs12_pages = NULL,
 	.nested_enable_evmcs = NULL,
+	.nested_get_evmcs_version = NULL,
 	.need_emulation_on_page_fault = vmx_need_emulation_on_page_fault,
 	.apic_init_signal_blocked = vmx_apic_init_signal_blocked,
 };

From patchwork Tue Aug 27 16:04:03 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Vitaly Kuznetsov <vkuznets@redhat.com>
X-Patchwork-Id: 11117241
Return-Path: <SRS0=KFyJ=WX=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 85C041395
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 27 Aug 2019 16:04:34 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id 6F58D21883
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 27 Aug 2019 16:04:34 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1730333AbfH0QEd (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 27 Aug 2019 12:04:33 -0400
Received: from mx1.redhat.com ([209.132.183.28]:37875 "EHLO mx1.redhat.com"
        rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
        id S1729903AbfH0QEL (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 27 Aug 2019 12:04:11 -0400
Received: from mail-wm1-f71.google.com (mail-wm1-f71.google.com
 [209.85.128.71])
        (using TLSv1.2 with cipher ECDHE-RSA-AES128-GCM-SHA256 (128/128 bits))
        (No client certificate requested)
        by mx1.redhat.com (Postfix) with ESMTPS id 8DBE77BDA5
        for <kvm@vger.kernel.org>; Tue, 27 Aug 2019 16:04:11 +0000 (UTC)
Received: by mail-wm1-f71.google.com with SMTP id r9so1243207wme.8
        for <kvm@vger.kernel.org>; Tue, 27 Aug 2019 09:04:11 -0700 (PDT)
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=/XqWagcDFu3I/3+ooG2WadK5AEABMMtW/Vv17jrsAvY=;
        b=lb6cOoFK1PNhO5UBVc6KKZDi+cTXA6GtXFkeJviFH+NHy5l0QnwJQM9WD8+OtqE/0b
         crZUc76EtHHA0NUn5JwwaWpl2ALKJeszFNlYuOtNTdFa+FvUFfv5cq/Md0G/u8Cf6DJg
         lnugfvh89Q13PXoob9EJ+i9j2+wqoVH1YNphAK2YIWmr4tYLpSDR9p0nNbiPOpuVRZ4W
         wsa4pSo1BvgwBCw2tRm7EVi8l5gu89du9ODagpd33zQK9Vt1jhBUuoZb1zQONq+I4mX1
         QaGf5KdQMNTr0gjCYIdNk5N/iwtQLS4mVKaAg4CckCGwsqFJ0BF0/mBSmAGliVKaTKYu
         Jqrg==
X-Gm-Message-State: APjAAAXncnRJbwMnjmrFAM9ECg+B0NEVu/DfLjtD/MM4KFNgPdm5BQZX
        H6jGt58gnGr2JlS1KjqTfUGv3jSxbPFdmtI74bD/vY6PNzonIqxxtYZ0TVzbqTCYbOoQzUIuAss
        Hfp8xrG+nqdMI
X-Received: by 2002:a7b:c157:: with SMTP id
 z23mr29337869wmi.104.1566921849783;
        Tue, 27 Aug 2019 09:04:09 -0700 (PDT)
X-Google-Smtp-Source: 
 APXvYqybdX7p/8QSoNaXF11fp+waAg3kYnqCagdBNshParsrqdkSV4cb0vJ1fCD53SJ5jbyMRhQ3aw==
X-Received: by 2002:a7b:c157:: with SMTP id
 z23mr29337838wmi.104.1566921849575;
        Tue, 27 Aug 2019 09:04:09 -0700 (PDT)
Received: from vitty.brq.redhat.com (nat-pool-brq-t.redhat.com.
 [213.175.37.10])
        by smtp.gmail.com with ESMTPSA id
 n8sm13461246wro.89.2019.08.27.09.04.08
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Tue, 27 Aug 2019 09:04:08 -0700 (PDT)
From: Vitaly Kuznetsov <vkuznets@redhat.com>
To: kvm@vger.kernel.org
Cc: linux-kernel@vger.kernel.org, Paolo Bonzini <pbonzini@redhat.com>,
	=?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Joerg Roedel <joro@8bytes.org>, Jim Mattson <jmattson@google.com>,
 Sean Christopherson <sean.j.christopherson@intel.com>,
 Roman Kagan <rkagan@virtuozzo.com>
Subject: [PATCH 2/3] KVM: x86: svm: remove unneeded nested_enable_evmcs() hook
Date: Tue, 27 Aug 2019 18:04:03 +0200
Message-Id: <20190827160404.14098-3-vkuznets@redhat.com>
X-Mailer: git-send-email 2.20.1
In-Reply-To: <20190827160404.14098-1-vkuznets@redhat.com>
References: <20190827160404.14098-1-vkuznets@redhat.com>
MIME-Version: 1.0
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Since commit 5158917c7b019 ("KVM: x86: nVMX: Allow nested_enable_evmcs to
be NULL") the code in x86.c is prepared to see nested_enable_evmcs being
NULL and in VMX case it actually is when nesting is disabled. Remove the
unneeded stub from SVM code.

Signed-off-by: Vitaly Kuznetsov <vkuznets@redhat.com>
Reviewed-by: Jim Mattson <jmattson@google.com>
---
 arch/x86/kvm/svm.c | 9 +--------
 1 file changed, 1 insertion(+), 8 deletions(-)

diff --git a/arch/x86/kvm/svm.c b/arch/x86/kvm/svm.c
index 40175c42f116..6d52c65d625b 100644
--- a/arch/x86/kvm/svm.c
+++ b/arch/x86/kvm/svm.c
@@ -7122,13 +7122,6 @@ static int svm_unregister_enc_region(struct kvm *kvm,
 	return ret;
 }
 
-static int nested_enable_evmcs(struct kvm_vcpu *vcpu,
-				   uint16_t *vmcs_version)
-{
-	/* Intel-only feature */
-	return -ENODEV;
-}
-
 static bool svm_need_emulation_on_page_fault(struct kvm_vcpu *vcpu)
 {
 	unsigned long cr4 = kvm_read_cr4(vcpu);
@@ -7337,7 +7330,7 @@ static struct kvm_x86_ops svm_x86_ops __ro_after_init = {
 	.mem_enc_reg_region = svm_register_enc_region,
 	.mem_enc_unreg_region = svm_unregister_enc_region,
 
-	.nested_enable_evmcs = nested_enable_evmcs,
+	.nested_enable_evmcs = NULL,
 	.nested_get_evmcs_version = NULL,
 
 	.need_emulation_on_page_fault = svm_need_emulation_on_page_fault,

From patchwork Tue Aug 27 16:04:04 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Vitaly Kuznetsov <vkuznets@redhat.com>
X-Patchwork-Id: 11117237
Return-Path: <SRS0=KFyJ=WX=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id E83DA1395
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 27 Aug 2019 16:04:18 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id D13882173E
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 27 Aug 2019 16:04:18 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1730243AbfH0QER (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 27 Aug 2019 12:04:17 -0400
Received: from mx1.redhat.com ([209.132.183.28]:49910 "EHLO mx1.redhat.com"
        rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
        id S1730182AbfH0QEN (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 27 Aug 2019 12:04:13 -0400
Received: from mail-wm1-f72.google.com (mail-wm1-f72.google.com
 [209.85.128.72])
        (using TLSv1.2 with cipher ECDHE-RSA-AES128-GCM-SHA256 (128/128 bits))
        (No client certificate requested)
        by mx1.redhat.com (Postfix) with ESMTPS id 0D80069061
        for <kvm@vger.kernel.org>; Tue, 27 Aug 2019 16:04:13 +0000 (UTC)
Received: by mail-wm1-f72.google.com with SMTP id c14so1248722wml.5
        for <kvm@vger.kernel.org>; Tue, 27 Aug 2019 09:04:12 -0700 (PDT)
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=ATtFby2HHOKxvIMlAxWSWxfBwEvUPTz5bp6RBKSRe98=;
        b=g7bAdPJBhcMS8/WMkPMfx+uWVHVEG3GyuiG+bbtYHNaoz8LXDBtjW0oKz8cCYnmhts
         LZJVbP5svUCljLGfYP8urSn0vhopJhRM9TgKZBbY9LbzpoIu2Z3Io33FBac9UeDR50i+
         l48ho8BPKkGxCyC2reIxYZFkp4jLOK+vQWgAEC/BzCewFFEpwIdO0lggfRkMr4/TdvPU
         /BjFgFUUtvMm53LWGNHiWdjH2xTh1Ai8EwzXki6kuonjXLItH8LGkfvmRZoluaVFmu35
         SsEsPYY45f2elOxFl2Kg0xCCeLhzPtDZb7iP30RpY4BET/gGGWZGY286AhXn9FGiDbPK
         rv3Q==
X-Gm-Message-State: APjAAAVecTC9jX0zOEFrIb7rmyfhZMfFdjcVpymOYQaFXUT+UrWwRDDm
        yv3rZVCb6qgfl7NN1ZAiPNWQQ7kPiTf7h+Ej0RTWV0Wike/h24cQrvMQ1Xi3qwXXExsC8kE0/ZS
        533zkAofB9lHY
X-Received: by 2002:a1c:b342:: with SMTP id c63mr29455687wmf.84.1566921851187;
        Tue, 27 Aug 2019 09:04:11 -0700 (PDT)
X-Google-Smtp-Source: 
 APXvYqw5+ZjwKbX+HfEYgJbnHxYapRoHIk9gLuCdB4RLCgMobGff6yZoB0Y/LUoJdbxRNnPysx3N+A==
X-Received: by 2002:a1c:b342:: with SMTP id c63mr29455658wmf.84.1566921850970;
        Tue, 27 Aug 2019 09:04:10 -0700 (PDT)
Received: from vitty.brq.redhat.com (nat-pool-brq-t.redhat.com.
 [213.175.37.10])
        by smtp.gmail.com with ESMTPSA id
 n8sm13461246wro.89.2019.08.27.09.04.09
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Tue, 27 Aug 2019 09:04:09 -0700 (PDT)
From: Vitaly Kuznetsov <vkuznets@redhat.com>
To: kvm@vger.kernel.org
Cc: linux-kernel@vger.kernel.org, Paolo Bonzini <pbonzini@redhat.com>,
	=?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Joerg Roedel <joro@8bytes.org>, Jim Mattson <jmattson@google.com>,
 Sean Christopherson <sean.j.christopherson@intel.com>,
 Roman Kagan <rkagan@virtuozzo.com>
Subject: [PATCH 3/3] KVM: x86: announce KVM_CAP_HYPERV_ENLIGHTENED_VMCS
 support only when it is available
Date: Tue, 27 Aug 2019 18:04:04 +0200
Message-Id: <20190827160404.14098-4-vkuznets@redhat.com>
X-Mailer: git-send-email 2.20.1
In-Reply-To: <20190827160404.14098-1-vkuznets@redhat.com>
References: <20190827160404.14098-1-vkuznets@redhat.com>
MIME-Version: 1.0
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

It was discovered that after commit 65efa61dc0d5 ("selftests: kvm: provide
common function to enable eVMCS") hyperv_cpuid selftest is failing on AMD.
The reason is that the commit changed _vcpu_ioctl() to vcpu_ioctl() in the
test and this one can't fail.

Instead of fixing the test is seems to make more sense to not announce
KVM_CAP_HYPERV_ENLIGHTENED_VMCS support if it is definitely missing
(on svm and in case kvm_intel.nested=0).

Signed-off-by: Vitaly Kuznetsov <vkuznets@redhat.com>
---
 arch/x86/kvm/x86.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index d1cd0fcff9e7..ef2e8b138300 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -3106,7 +3106,6 @@ int kvm_vm_ioctl_check_extension(struct kvm *kvm, long ext)
 	case KVM_CAP_HYPERV_EVENTFD:
 	case KVM_CAP_HYPERV_TLBFLUSH:
 	case KVM_CAP_HYPERV_SEND_IPI:
-	case KVM_CAP_HYPERV_ENLIGHTENED_VMCS:
 	case KVM_CAP_HYPERV_CPUID:
 	case KVM_CAP_PCI_SEGMENT:
 	case KVM_CAP_DEBUGREGS:
@@ -3183,6 +3182,8 @@ int kvm_vm_ioctl_check_extension(struct kvm *kvm, long ext)
 		r = kvm_x86_ops->get_nested_state ?
 			kvm_x86_ops->get_nested_state(NULL, NULL, 0) : 0;
 		break;
+	case KVM_CAP_HYPERV_ENLIGHTENED_VMCS:
+		r = kvm_x86_ops->nested_enable_evmcs != NULL;
 	default:
 		break;
 	}
